### ISSUE TITLE
Failed to load AppArmor policy  

### ISSUE THREAD

**Issue by pushou** [canonical/multipass#1216]
hi everyone,

OS ubuntu 18.04 
❯ uname -a                                                                      ─╯
Linux ..Precision-3520 5.0.21-050021-generic #201906040731 SMP Tue Jun 4 07:33:07 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
❯ multipass version                                                   ─╯
multipass  0.10.0-dev.116+ga77aa05
multipassd 0.10.0-dev.116+ga77aa05
         

multipass worked perfectly until yesterday.

```
multipass launch bionic -n u18cc1                                                                                  ─╯
launch failed: Failed to load AppArmor policy                                   
#include <tunables/global>
profile multipass.dnsmasq flags=(attach_disconnected) {
  #include <abstractions/base>
  #include <abstractions/nameservice>

  capability chown,
  capability net_bind_service,
  capability setgid,
  capability setuid,
  capability dac_override,
  capability dac_read_search,
  capability net_admin,         # for DHCP server
  capability net_raw,           # for DHCP server ping checks
  network inet raw,
  network inet6 raw,

  # Allow multipassd send dnsmasq signals
  signal (receive) peer=snap.multipass.multipassd,

  # access to iface mtu needed for Router Advertisement messages in IPv6
  # Neighbor Discovery protocol (RFC 2461)
  @{PROC}/sys/net/ipv6/conf/*/mtu r,

  # binary and its libs
  /snap/multipass/1434/usr/sbin/dnsmasq ixr,
  /snap/multipass/1434/{usr/,}lib/@{multiarch}/{,**/}*.so* rm,

  # CLASSIC ONLY: need to specify required libs from core snap
  /snap/core18/*/{,usr/}lib/@{multiarch}/{,**/}*.so* rm,

  /var/snap/multipass/common/data/multipassd/network/dnsmasq.leases rw,           # Leases file
  /var/snap/multipass/common/data/multipassd/network/dnsmasq.hosts r,             # Hosts file

  /var/snap/multipass/common/data/multipassd/network/dnsmasq.pid w,     # pid file
}
    : errno=1 ()

❯             
grep apparmor  strace.log                                                     ─╯
newfstatat(AT_FDCWD, "/etc/apparmor.d/cache", {st_mode=S_IFDIR|0755, st_size=4096, ...}, 0) = 0
openat(AT_FDCWD, "/sys/kernel/security/apparmor/features", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/etc/apparmor.d/cache", {st_mode=S_IFDIR|0755, st_size=4096, ...}, 0) = 0
openat(AT_FDCWD, "/sys/kernel/security/apparmor/features", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/usr/local/sbin/apparmor_parser", 0xc420320038, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/usr/local/bin/apparmor_parser", 0xc420320108, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/usr/sbin/apparmor_parser", 0xc4203201d8, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/usr/bin/apparmor_parser", 0xc4203202a8, 0) = -1 ENOENT (No such file or directory)
newfstatat(AT_FDCWD, "/sbin/apparmor_parser", {st_mode=S_IFREG|0755, st_size=1472232, ...}, 0) = 0
newfstatat(AT_FDCWD, "/sbin/apparmor_parser", {st_mode=S_IFREG|0755, st_size=1472232, ...}, 0) = 0
stat("/sys/kernel/security/apparmor", 0x7ffe49af7870) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/sys/module/apparmor/parameters/enabled", O_RDONLY) = 4
newfstatat(AT_FDCWD, "/etc/apparmor.d/cache", {st_mode=S_IFDIR|0755, st_size=4096, ...}, 0) = 0
openat(AT_FDCWD, "/snap/multipass/1434/bin/../lib/tls/haswell/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../lib/tls/haswell/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../lib/tls/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../lib/tls/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../lib/haswell/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../lib/haswell/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../lib/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../lib/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../usr/lib/x86_64-linux-gnu/tls/haswell/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../usr/lib/x86_64-linux-gnu/tls/haswell/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../usr/lib/x86_64-linux-gnu/tls/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../usr/lib/x86_64-linux-gnu/tls/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../usr/lib/x86_64-linux-gnu/haswell/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../usr/lib/x86_64-linux-gnu/haswell/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../usr/lib/x86_64-linux-gnu/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/multipass/1434/bin/../usr/lib/x86_64-linux-gnu/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/core18/current/lib/x86_64-linux-gnu/tls/haswell/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/core18/current/lib/x86_64-linux-gnu/tls/haswell/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/core18/current/lib/x86_64-linux-gnu/tls/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/core18/current/lib/x86_64-linux-gnu/tls/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/core18/current/lib/x86_64-linux-gnu/haswell/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/core18/current/lib/x86_64-linux-gnu/haswell/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/core18/current/lib/x86_64-linux-gnu/x86_64/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
openat(AT_FDCWD, "/snap/core18/current/lib/x86_64-linux-gnu/libapparmor.so.1", O_RDONLY|O_CLOEXEC) = 3
```

Using libvirt backend multipass works and I can run VM.

I try to re-install , flush snap dir , build from source  without  any success
thanks for you help

**Comment by Saviq** [comment]
Hi @pushou, sorry about this, we're preparing to enable strict mode in Multipass (#1074), so we had to add some AppArmor profiles.

The kernel you're using doesn't seem to be an Ubuntu kernel, though? It must not be supporting the features of AppArmor we rely on.

Can you please try switching to an Ubuntu kernel, and we'll see how to degrade gracefully in this case.

**Comment by pushou** [comment]
Thanks Saviq for your answers
my kernel is an hwe kernel . it is an ubuntu kernel isn'it ?
but  downgrading to the previous installed version is Ok.  
the new version is the problem

**Comment by Saviq** [comment]
@pushou that's a mainline kernel, not the HWE one:

https://packages.ubuntu.com/search?keywords=linux-generic-hwe-18.04

That said, we need to support mainline, too - stay tuned.

**Comment by Saviq** [comment]
Hi @pushou would you be able to try the fix in #1218? Just issue this:
```
snap refresh multipass --channel edge/pr1218
```

Thanks!

**Comment by pushou** [comment]
Thanks all works fine
